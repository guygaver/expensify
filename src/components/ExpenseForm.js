import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        
        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense ? ((props.expense.amount / 100) * 100).toString() : "",
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ''
        }; 
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState({
            description
        })
    };

    onAmountChange = (e) => {
        const amount = e.target.value;

        if ( !amount || amount.match(/^\d*(\.\d{0,2})?$/) ) {
            this.setState({
                amount
            })
        }
    };

    onDateChange = (createdAt) => {

        if ( createdAt ) {
            this.setState({createdAt})
        }
    };

    onNoteChange = (e) => {
        const note = e.target.value;

        this.setState({
            note
        })
    };

    onSubmit = (e) => {
        e.preventDefault();

        if ( !this.state.description || !this.state.amount ) {
            this.setState({error: 'Please provide description and amount'})
        } else {
            this.setState({error: ''});
            
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note,
            })
        }
    };

    render() {
        return (
            <form onSubmit={this.onSubmit} className="form">
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="Description"
                    autoFocus={true}
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    className="text-input"
                />

                <input
                    type="text"
                    placeholder="Amount"
                    onChange={this.onAmountChange}
                    value={this.state.amount}
                    className="text-input"
                />

                <SingleDatePicker
                    id={"single-date-picker"}
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calenderFocused}
                    onFocusChange={({focused}) => this.setState({calenderFocused: focused})}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />

                <textarea
                    value={this.state.note}
                    placeholder="Add a note for your expense (optional)"
                    onChange={this.onNoteChange}
                    className="text-area"
                >
                    
                </textarea>

                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        )
    }
}
