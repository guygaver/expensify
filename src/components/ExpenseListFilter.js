import React from 'react'
import {connect} from 'react-redux'
import {DateRangePicker} from 'react-dates'
import {setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from "../actions/filters";

class ExpenseListFilter extends React.Component {
    
    state = {
        calenderFocused: null
    };

    onDatesChange = ( {startDate, endDate} ) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    };

    onFocusChange = (calenderFocused) => {this.setState({calenderFocused})};
    
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type="text"
                            value={this.props.filters.text}
                            onChange={(e) => {
                                this.props.dispatch(setTextFilter(e.target.value))
                            }}/>
                    </div>
                    <div className="input-group__item">
                        <select
                            value={this.props.filters.sortBy}
                            onChange={(e) => {
                                const value = e.target.value;

                                if ( value === 'date' ) {
                                    this.props.dispatch(sortByDate())
                                } else if ( value === 'amount' ) {
                                    this.props.dispatch(sortByAmount())
                                }
                            }}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calenderFocused}
                            onFocusChange={this.onFocusChange}
                            startDateId="range-start"
                            endDateId="range-end"
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </div>
                </div>
            </div>
        )
        
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};

export default connect(mapStateToProps)(ExpenseListFilter);