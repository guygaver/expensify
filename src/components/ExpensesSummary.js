import React from 'react'
import expensesTotal from './../selectors/expenses-total'
import selectExpenses from './../selectors/expenses'
import {connect} from "react-redux";
import numeral from 'numeraljs';

export const ExpensesSummary = ({count, total}) => (
    <div>
        <h1>Viewing {count} expense/s totaling {numeral(total).format('$0,0.00')}</h1>
    </div>
);

const mapStateToProps = (state) => {
    let visibleExpenses = selectExpenses(state.expenses, state.filters);
    
    return {
        total: expensesTotal(visibleExpenses),
        count: visibleExpenses.length
    }
};

export default connect(mapStateToProps)(ExpensesSummary);