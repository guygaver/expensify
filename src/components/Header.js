import React from 'react'
import {Link} from 'react-router-dom'
import {connect} from 'react-redux'
import { startLogout }from './../actions/auth'

export const Header = ({ logOut }) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/dashboard">
                    <h1>Expensify</h1>
                </Link>
                <button className="button button--link" onClick={logOut}>Logout</button>
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    logOut: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);