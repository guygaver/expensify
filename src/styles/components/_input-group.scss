.input-group {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  margin-bottom: $m-size;
  
  @media (min-width: 100px) {
    flex-direction: row;
    margin-bottom: $l-size;
  }
}

.input-group__item {
  margin-bottom: $s-size;

  @media (min-width: 100px) {
    margin: 0 $s-size 0 0 ;
  }
}