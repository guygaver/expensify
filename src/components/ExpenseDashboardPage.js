import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilter from './ExpenseListFilter'
import ExpensesSummary from './ExpensesSummary'

export const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilter />
        <ExpenseList />
    </div>
);


export default ExpenseDashboardPage;