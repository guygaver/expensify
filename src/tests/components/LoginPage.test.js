import React from 'react'
import {shallow} from 'enzyme'
import {LoginPage} from './../../components/LoginPage'

test('it should render LoginPage component', () => {
    const wrapper = shallow(<LoginPage />);
    expect(wrapper).toMatchSnapshot();
});

test('should call logOut on button click', () => {
    const logIn = jest.fn();
    const wrapper = shallow(<LoginPage startLogin={logIn}/>);
    wrapper.find('button').simulate('click');

    expect(logIn).toHaveBeenCalled();
});