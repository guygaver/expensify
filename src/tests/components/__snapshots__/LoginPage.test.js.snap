// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render LoginPage component 1`] = `
<div>
  <button>
    Login
  </button>
</div>
`;
